### ISSUE TITLE
Conan2: How to set options of dependent packages?

### ISSUE THREAD

**Issue by gouriano** [conan-io/conan#13383]
### What is your question?

For example, my package requires openssl.

On command line, I can specify

      conan install . --options=openssl/*:shared=True

In conanfile.txt:

    [options]
    openssl*:shared=True

With Conan2, how to do that in conanfile.py? 
In conan1 the following worked:

    self.options["openssl"].shared = True

In Conan2, this results in an exception.
Sorry, I cannot find any information in the documentation.
Thanks

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @gouriano 

Thanks for your question. In Conan 2.0, the syntax of the ``conanfile.py`` would be similar to the one in the conanfile.txt, and a pattern is needed, so try:
```python
     self.options["openssl/*"].shared = True
```
Still, the best way to define variations over options is not ``conanfiles``, but profiles or command line. When different ``conanfile`` in a dependency graph define different values for dependencies options, they will compete, and basically the first one to arrive wins and sets the priority. While in profiles, the definition always has higher priority, so it is more clear that it is being defined.

**Comment by SpaceIm** [comment]
How to achieve v1/v2 cross-compability in recipes which have to force dependencies options when it's an explicit requirement to have such value? (In conan center, we force such options in `configure()` then raise in `validate()` if option has been overridden by another recipe or downstream, very common pattern in dependencies of glib).

**Comment by memsharded** [comment]
Still the same, define things in ``configure()`` and raise in ``validate()``

**Comment by gouriano** [comment]
In my case, it is

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=17
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=192
os=Windows
[options]
ncbi-cxx-toolkit-core/*:with_targets=xncbi
[conf]


Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=17
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=192
os=Windows
[conf]



======== Computing dependency graph ========
Package requires  bzip2/1.0.8
Package requires  pcre/8.45
Package requires  libuv/1.44.2
Package requires  libnghttp2/1.51.0
Package requires  zstd/1.5.2
Package requires  lzo/2.10
Package requires  zlib/1.2.13
Graph root
    cli
Requirements
    bzip2/1.0.8#23cc41191fd58460207c5db046033507 - Cache
    libnghttp2/1.51.0#1757a7250ffc076950acc0f16c3e8e6b - Cache
    libuv/1.44.2#656c5f28ec978f7b8d4fcc55916dcb72 - Cache
    lzo/2.10#850936b57f92160152bd9385b1770a69 - Cache
    ncbi-cxx-toolkit-core/0.0.0#cd2d185538f1aee6c5bdd4801f4f753a - Cache
    pcre/8.45#563d050c69aaa0a61be46bb09ba5c23c - Cache
    zlib/1.2.13#13c96f538b52e1600c40b88994de240f - Cache
    zstd/1.5.2#74cd46b9525f6d1af311660ef17ca48c - Cache

======== Computing necessary packages ========
Traceback (most recent call last):
  File "g:\conan\venv_conan2\lib\site-packages\conan\cli\cli.py", line 264, in main
    cli.run(args)
  File "g:\conan\venv_conan2\lib\site-packages\conan\cli\cli.py", line 167, in run
    command.run(self._conan_api, self._commands[command_argument].parser, args[0][1:])
  File "g:\conan\venv_conan2\lib\site-packages\conan\cli\command.py", line 130, in run
    info = self._method(conan_api, parser, *args)
  File "g:\conan\venv_conan2\lib\site-packages\conan\cli\commands\create.py", line 78, in create
    update=args.update, lockfile=lockfile)
  File "g:\conan\venv_conan2\lib\site-packages\conan\api\subapi\graph.py", line 190, in analyze_binaries
    binaries_analyzer.evaluate_graph(graph, build_mode, lockfile, remotes, update)
  File "g:\conan\venv_conan2\lib\site-packages\conans\client\graph\graph_binaries.py", line 324, in evaluate_graph
    self._evaluate_package_id(node)
  File "g:\conan\venv_conan2\lib\site-packages\conans\client\graph\graph_binaries.py", line 292, in _evaluate_package_id
    compute_package_id(node, self._cache.new_config)  # TODO: revise compute_package_id()
  File "g:\conan\venv_conan2\lib\site-packages\conans\client\graph\compute_pid.py", line 44, in compute_package_id
    options=conanfile.options.copy_conaninfo_options(),
  File "g:\conan\venv_conan2\lib\site-packages\conans\model\options.py", line 341, in copy_conaninfo_options
    assert not self._deps_package_options
AssertionError

ERROR:

**Comment by gouriano** [comment]
In my case it is                          self.options["libnghttp2/*"].shared = False

**Comment by memsharded** [comment]
```
assert not self._deps_package_options
```
This was happening when options manipulation is done in ``requirements()`` method, which is forbidden (or ignored in 1.X). A better error message was introduced in Conan 2.0.1, which Conan version are you using?

**Comment by gouriano** [comment]
I use v2.0.0, and yes, I do this in requirements() method. (With Conan1, I did the same). So, you say I should do it in configure()?

**Comment by memsharded** [comment]
Yes, please upgrade to 2.0.1 and you will get a better error message.

yes, in Conan 1.X that option was most likely being ignored. It should be done in ``configure()`` method.

**Comment by SpaceIm** [comment]
> Still the same, define things in `configure()` and raise in `validate()`

Sorry I don't understand the answer. If syntax is different, how to easily achive cross-compability? My understanding is that CCI recipes doing self.options["openssl"].shared = True currently are broken with conan v2 client (not realy broken, but it's not honored). If we write self.options["openssl/*"].shared = True, they are broken for conan v1 client. That's my point.

**Comment by gouriano** [comment]
That is, I define _openssl_ options even before I _require_ it. Counterintuitive, in my opinion, but it works. Great. Thank you.

**Comment by memsharded** [comment]
`` self.options["openssl*"].shared = True`` syntax works for 1.X, it should be good enough to allow the migration

**Comment by SpaceIm** [comment]
> ` self.options["openssl*"].shared = True` syntax works for 1.X, it should be good enough to allow the migration

Seems fragile since it may change shared option of another recipe opensslfoo. I don't see any robust solution except through conan_version branching.

**Comment by memsharded** [comment]
Does ``opensslfoo`` exists? is ``opensslfoo`` a dependency of the current recipe? 

I think that we can be very sure that this will not happen, so it doesn't seem fragile to me, but a very pragmatic and reasonable approach.

**Comment by SpaceIm** [comment]
It's just an hypothetical example. There are 1500 recipes in CCI and nobody knows which new recipe name may be introduced, nor if some dependency may not change its own dependency graph. I don't trust a `self.options["foo*"]`, it's not future proof.

**Comment by memsharded** [comment]
That is my believe, it is very hypothetical, not real. There are very few overlapping names in ConanCenter, and the names of the dependencies are very controlled, it is not like overlapping conflicting dependencies will suddenly appear in your transitive dependencies, because at the end of the day the ``requires = "zlib/1.2.13"`` is hardcoded, and it will not change spontaneously to ``zlib-ng`` or something like that.

Plus it doesn't need to be future-proof forever, when Conan 1 support is dropped in ConanCenter, we will be able to move to the ``[zlib/*]`` safe syntax. No need to add a lot of clutter to the recipes for some extremely unlikely collision.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/13467 will try to add back in 2.0.3 the syntax with ``self.options["mydep"].myoption`` without explicit "*"

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/13467 merged, it will be in 2.0.3
